src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11761,54 +11812,39 @@ private bool CanStartConsequenceExpression()
                 return this.CurrentToken.Kind != SyntaxKind.ColonToken;
             }
 
-            // Anything else is just a normal ? and indicates the start of a ternary expression.
-            return false;
-        }
-
-        internal ExpressionSyntax ParseConsequenceSyntax()
-        {
-            Debug.Assert(this.CurrentToken.Kind is SyntaxKind.DotToken or SyntaxKind.OpenBracketToken);
-            ExpressionSyntax expr = this.CurrentToken.Kind switch
+            ExpressionSyntax parseWhenNotNull(ExpressionSyntax expr)
             {
-                SyntaxKind.DotToken => _syntaxFactory.MemberBindingExpression(this.EatToken(), this.ParseSimpleName(NameOptions.InExpression)),
-                SyntaxKind.OpenBracketToken => _syntaxFactory.ElementBindingExpression(this.ParseBracketedArgumentList()),
-                _ => throw ExceptionUtilities.Unreachable(),
```

Comment:
no need for this switch anymore.  we just inline this to the point that we check for `.` and `[`